### PR TITLE
pillar/devicenetwork: fix goroutine leak

### PR DESCRIPTION
**Pull request by christoph-zededa**
introduced in https://github.com/lf-edge/eve/commit/85733729e0606f1a0637b1853f34f1fb5c1a3168

without this fix, the goroutine is blocked by trying to send
into a channel that has no receiver

### CODE REVIEW

**Review by milan-zededa — APPROVED** [review]

**Review by OhmSpectator — APPROVED** [review]
I prefer just to believe it works... I tried and failed to understand the async and channel magic in the function. We use one channel as a semaphore, another - for signalling an IP and the third for checking to quit...

**Comment by OhmSpectator** [comment]
@christoph-zededa, I see you moved the PR into the draft state. Any concerns there?

**Comment by christoph-zededa** [comment]
> @christoph-zededa, I see you moved the PR into the draft state. Any concerns there?

There were some more leaks ...

File: pkg/pillar/devicenetwork/dns.go
Comment (OhmSpectator) [inline pkg/pillar/devicenetwork/dns.go]: Did you see this leak in the `sigusr1` log file we investigated earlier?

**Comment by OhmSpectator** [comment]
> > I prefer just to believe it works... I tried and failed to understand the async and channel magic in the function. We use one channel as a semaphore, another - for signalling an IP and the third for checking to quit...
> 
> I heard you want more: https://github.com/lf-edge/eve/pull/4409/files#diff-73f9b01c33a31700345aa9b1f29289dfdb697a5593a9585cf68443155aa12e50R169

😨😨😨

File: pkg/pillar/devicenetwork/dns.go
Comment (OhmSpectator) [inline pkg/pillar/devicenetwork/dns.go:169]: Is it a fix for a leak or a prerequisite for a test? Are you sure it's in the right commit?

**Review by OhmSpectator — APPROVED** [review]
Can we use the same leak-detecting lib for the other components as well?

**Comment by OhmSpectator** [comment]
@christoph-zededa, if you find and fix more leaks, could you please drop here information about which versions are affected by them? We need this information to understand where to backport (or just inform users who use these versions). Also, it may make sense to make a commit per fix to make it easier to track the backports.

File: pkg/pillar/devicenetwork/dns.go
Comment (milan-zededa) [inline pkg/pillar/devicenetwork/dns.go:178]: Why not just use `<-wgChan` here?

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:169]: It is needed for test, because otherwise the following goroutine would still persist when the test already finished and then it would get detected by the leak detector.

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go]: I see it now.
There are goroutines that are stuck in line 164 waiting for `wg.Wait()` and because `quit` was not closed in all cases, not all `wg.Done()` were called.

**Comment by christoph-zededa** [comment]
> @christoph-zededa, if you find and fix more leaks, could you please drop here information about which versions are affected by them? We need this information to understand where to backport (or just inform users who use these versions). Also, it may make sense to make a commit per fix to make it easier to track the backports.

They were all introduced with https://github.com/lf-edge/eve/commit/85733729e0606f1a0637b1853f34f1fb5c1a3168 .

File: pkg/pillar/devicenetwork/dns.go
Comment (OhmSpectator) [reply pkg/pillar/devicenetwork/dns.go]: Ok, I see. 19170 occurrences.

**Comment by OhmSpectator** [comment]
> > @christoph-zededa, if you find and fix more leaks, could you please drop here information about which versions are affected by them? We need this information to understand where to backport (or just inform users who use these versions). Also, it may make sense to make a commit per fix to make it easier to track the backports.
> 
> They were all introduced with [8573372](https://github.com/lf-edge/eve/commit/85733729e0606f1a0637b1853f34f1fb5c1a3168) .

Got it, thanks!

**Review by eriknordmark — APPROVED** [review]
LGTM.

Did the added test logic detect the failiure in the old code?
(Running that test on the old code would be a useful check.)

**Review by OhmSpectator — APPROVED** [review]

**Comment by christoph-zededa** [comment]
> LGTM.
> 
> Did the added test logic detect the failiure in the old code? (Running that test on the old code would be a useful check.)

Yes, that's how I found all the additional problems; the original fix was only:
```
diff --git a/pkg/pillar/devicenetwork/dns.go b/pkg/pillar/devicenetwork/dns.go
index 7f31eb11e..80f49be55 100644
--- a/pkg/pillar/devicenetwork/dns.go
+++ b/pkg/pillar/devicenetwork/dns.go
@@ -150,7 +150,10 @@ func ResolveWithPortsLambda(domain string,
                                                errs = append(errs, err)
                                        }
                                        if response != nil {
-                                               resolvedIPsChan <- response
+                                               select {
+                                               case <-resolvedIPsChan:
+                                               case resolvedIPsChan <- response:
+                                               }
                                        }
                                        <-work
                                        wg.Done()
@@ -183,6 +186,7 @@ func ResolveWithPortsLambda(domain string,
                }
                return responses, errs
        case ip := <-resolvedIPsChan:
+               close(resolvedIPsChan)
                close(quit)
                return ip, nil
        }
```

**Comment by christoph-zededa** [comment]
No idea what problems eden has, is it because of Halloween?

Locally I could start eve and onboard it to the commercial controller.

**Comment by OhmSpectator** [comment]
> No idea what problems eden has, is it because of Halloween? 👻
> 
> Locally I could start eve and onboard it to the commercial controller.

All the tests show the same error:
```
time="2024-10-31T20:25:08Z" level=info msg="Adam waiting for EVE registration (14) of (20)"
time="2024-10-31T20:25:28Z" level=info msg="Adam waiting for EVE registration (15) of (20)"
time="2024-10-31T20:25:48Z" level=info msg="Adam waiting for EVE registration (16) of (20)"
time="2024-10-31T20:26:08Z" level=info msg="Adam waiting for EVE registration (17) of (20)"
time="2024-10-31T20:26:28Z" level=info msg="Adam waiting for EVE registration (18) of (20)"
time="2024-10-31T20:26:48Z" level=info msg="Adam waiting for EVE registration (19) of (20)"
time="2024-10-31T20:27:08Z" level=fatal msg="Eve onboard failed: error onboarding onboarding timeout. You may try to run 'eden eve onboard' command again in several minutes. If not successful see logs of adam/eve"
```

Some glitch, I hope. I am restarting the tests.

File: pkg/pillar/devicenetwork/dns.go
Comment (uncleDecart) [inline pkg/pillar/devicenetwork/dns.go:132]: len(wg) == countDNSRequests, so technically we don't need it.

File: pkg/pillar/devicenetwork/dns.go
Comment (uncleDecart) [inline pkg/pillar/devicenetwork/dns.go:154]: we can send errors through channel to avoid mutex

**Comment by OhmSpectator** [comment]
Tests restart does not help =\

File: pkg/pillar/devicenetwork/dns.go
Comment (uncleDecart) [inline pkg/pillar/devicenetwork/dns.go:178]: Okay, one thing I don't quite get is if we want any resolved IP addr, or all resolved IP addrs. Seems to me that we return first successful resolve. Because if we were to send all of the possible options, wrapping wg.Wait() in goroutine wouldn't make sense, we do wg.Wait() in function, block till we check all the options and then deal with results, because select is also blocking. 

In any of case I'd suggest we extract lookup function (that two loops with DNS and IPs) and write something like this:


```go

errs := []error
resolvedIPCh, errCh, outOfOptions := lookupIP(/* ... */)
select {
  case ip := <- resolvedIP:
     return ip, nil
  case <-outOfOptions:
     ips, err := net.LookupIP(domain)
     if err != nil {
     }
     // ...
  case err := <- errCh:
     errs.append(err)

} 

```

I think this way it's way more straightforward, I can write full snippet if you think it's a good approach @christoph-zededa

File: pkg/pillar/devicenetwork/dns_test.go
Comment (uncleDecart) [inline pkg/pillar/devicenetwork/dns_test.go:165]: :clap: :clap: :clap: 
leak sanitizer is a good thing, can we enable it test-wide?

**Review by uncleDecart — COMMENTED** [review]
Some small suggestions on restructuring code

File: pkg/pillar/devicenetwork/dns_test.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns_test.go:165]: You mean other tests?
I fear we will find more leaks and also more false-positives (like this opencensus thing).

Also it cannot run with `t.Parallel()` and needs a `TestMain(...)`.

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:132]: `wg` is not an array or a string, so I am not sure what `len(wg)` should do

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:154]: not sure if that is a gain, because then in the end I have to start another goroutine to receive them and append them to the array, or make the channel big enough to store them in there and then convert them to an array

File: pkg/pillar/devicenetwork/dns_test.go
Comment (OhmSpectator) [reply pkg/pillar/devicenetwork/dns_test.go:165]: > I fear we will find more leaks

😅😅😅
I think that's exactly what we need: find the leaks before users face the consequences of them =) But not within the scope of this PR.

**Review by OhmSpectator — APPROVED** [review]

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:178]: I am curious what that would look like.
I think breaking it up, would make it more readable.

File: pkg/pillar/devicenetwork/dns.go
Comment (cperakis) [inline pkg/pillar/devicenetwork/dns.go:159]: Its a great fix.

However, i would add add a timeout to the select statement to prevent the goroutine from blocking indefinitely when trying to send the DNS resolution response to the resolvedIPsChan channel. This will prevent from similar deadlocks in the future.

This timeout could be added on other similar code blocks.
```go
						case <-quit:
						case resolvedIPsChan <- response:
                                                case <-time.After(timeoutDuration):
                                                        return
``` 

Your opinion?

**Comment by cperakis** [comment]
@christoph-zededa @OhmSpectator @eriknordmark I think we should review how frequently we invoke this functionality to update the DNS resolution for the Ethernet interfaces' IP addresses.

In my opinion, the DNS names are unlikely to change often—probably only when the IP addresses of the interfaces themselves change. Therefore, we might not need to perform updates so frequently.

Also we could do that only if the IP of the ethernet interfaces will change.

What are your thoughts?

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:159]: We would need this for every `select` and for the `wg.Add(1)`, wouldn't we?

Let's see what will come out of https://github.com/lf-edge/eve/pull/4409#discussion_r1825637446 and perhaps that makes it easier ...

**Comment by christoph-zededa** [comment]
> @christoph-zededa @OhmSpectator @eriknordmark I think we should review how frequently we invoke this 

CC @milan-zededa 

> functionality to update the DNS resolution for the Ethernet interfaces' IP addresses.
> 
> In my opinion, the DNS names are unlikely to change often—probably only when the IP addresses of the interfaces themselves change. Therefore, we might not need to perform updates so frequently.
> 
> Also we could do that only if the IP of the ethernet interfaces will change.
> 
> What are your thoughts?

It is already cached here: https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/nim/resolvercache.go#L108

I guess we do not cache if resolving the name did not succeed and then it gets called again and again.

**Comment by uncleDecart** [comment]
> I guess we do not cache if resolving the name did not succeed and then it gets called again and again.

Yes, so calls are not that often

File: pkg/pillar/devicenetwork/dns.go
Comment (uncleDecart) [reply pkg/pillar/devicenetwork/dns.go:178]: Hold my Apfelschorle...

**Comment by OhmSpectator** [comment]
So, are we waiting for a new suggestion from @uncleDecart ?

**Comment by christoph-zededa** [comment]
> So, are we waiting for a new suggestion from @uncleDecart ?

IMO we don't have to wait; we can always create another PR.

**Review by OhmSpectator — APPROVED** [review]

File: pkg/pillar/devicenetwork/dns.go
Comment (mzampetakis) [inline pkg/pillar/devicenetwork/dns.go:174]: `resolvedIPsChan` and `work` channels are not closed at all. We could have a defer func to close them too.

File: pkg/pillar/devicenetwork/dns.go
Comment (mzampetakis) [reply pkg/pillar/devicenetwork/dns.go:154]: We could also use [sync.map[(https://pkg.go.dev/sync#Map) but I think Mx is fine for this case.

File: pkg/pillar/devicenetwork/dns.go
Comment (mzampetakis) [inline pkg/pillar/devicenetwork/dns.go:137]: I think that the following lines (
```
select {
    case work <- struct{}{}:
	// if writable, means less than dnsMaxParallelRequests goroutines are currently running
}
```
)
could be simplified to
```
work <- struct{}{}
```

**Comment by uncleDecart** [comment]
@christoph-zededa how do you like [this](https://gist.github.com/uncleDecart/09e3f551f16c7d81fa14acab519adff1)? I think that's easier to read, if it's good, I'll come up with some tests for it and we can include this solution

**Comment by christoph-zededa** [comment]
> @christoph-zededa how do you like [this](https://gist.github.com/uncleDecart/09e3f551f16c7d81fa14acab519adff1)? I think that's easier to read, if it's good, I'll come up with some tests for it and we can include this solution

@uncleDecart I agree it makes it easier to read. Thx.

**Comment by uncleDecart** [comment]
> Just create a PR into my PR.

So that you could review my PR while I review yours? Sure. I'll think about the tests and then open a PR 

> That's a new feature and not a bug fix, isn't it?

Agreed, I wouldn't go down that path, let's fix the bug

**Comment by OhmSpectator** [comment]
Mmm, guys, I would say, first of all, we need to fix the very specific bug in the scope of this PR. Time is still matters in this case. Refactoring and readability improvements - we can do it in dedicated PRs.

**Comment by uncleDecart** [comment]
> Mmm, guys, I would say, first of all, we need to fix the very specific bug in the scope of this PR. Time is still matters in this case. Refactoring and readability improvements - we can do it in dedicated PRs.

Then let's merge this PR @OhmSpectator and then I'll open mine

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:177]: They get gc-ed, but  I added it explicitly.

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:141]: done

File: pkg/pillar/devicenetwork/dns.go
Comment (eriknordmark) [reply pkg/pillar/devicenetwork/dns.go:162]: I think @cperakis is correct and that this is the root cause of the issue (especially since this PR didn't fix the issue). The channel is unbuffered hence this will block sending on resolvedIPsChan when is more than one response (assuming the channel reader goes away after the first response).

Comment (milan-zededa) [reply pkg/pillar/devicenetwork/dns.go:162]: But when the reader goes away, the `quit` channel is closed in the defer function and this should unblock this go routine.

Comment (eriknordmark) [reply pkg/pillar/devicenetwork/dns.go:162]: That doesn't follow normal golang patters - select across a send and a receive on channels is not something I've seen elsewhere.

Comment (OhmSpectator) [reply pkg/pillar/devicenetwork/dns.go:162]: > especially since this PR didn't fix the issue

it's because we tested the version without a fix, by my mistake...

Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:162]: Do you mean I should do instead:
```suggestion
						select {
						case resolvedIPsChan <- response:
						default:
						}
```

to make it non-blocking?

Comment (milan-zededa) [reply pkg/pillar/devicenetwork/dns.go:162]: > That doesn't follow normal golang patters - select across a send and a receive on channels is not something I've seen elsewhere.

It would be nicer to use context.Context with cancel, but actually `ctx.Done()` returns channel so select across send and receive is a common pattern.

Comment (uncleDecart) [reply pkg/pillar/devicenetwork/dns.go:161]: @milan-zededa how do you feel about [this](https://gist.github.com/uncleDecart/09e3f551f16c7d81fa14acab519adff1) approach?

Comment (uncleDecart) [reply pkg/pillar/devicenetwork/dns.go:161]: if we want to add timeout in that case we can do it quite easily. Just create a context with timeout :). I'm starting to think it'll make sense to do those changes now

Comment (eriknordmark) [reply pkg/pillar/devicenetwork/dns.go:161]: > Do you mean I should do instead:
> 
> to make it non-blocking?

That would ensure that it would never block. But it also doesn't wait for the receiver to try to read which AFAIU means that you might get nothing on the channel (that can be avoided by having a length of 1 for the channel).

But if the fix works then I think we should take it and move forward - maybe with some additional comments in the code but no code changes.

**Comment by christoph-zededa** [comment]
Test failed, because it also needs a:
```
-       resolvedIPsChan := make(chan []DNSResponse)
+       resolvedIPsChan := make(chan []DNSResponse, 1)
```

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:161]: Yes, that's why the `go test` sometimes failed.
(`go test -count 500 -failfast -race -v . -run=TestResolveWithPortsLambda$` to reproduce)

Making the channel of size `1` fixed it.

Comment (milan-zededa) [reply pkg/pillar/devicenetwork/dns.go:161]: @uncleDecart I think you should not need line 91: `cancel() // Cancel remeaning goroutines`
The context is canceled by defer of `resolveDNS`. 
Regarding timeout: we already have 30 seconds timeout inside `resolve()`. So I'm not sure if we want/need to have a global timeout for the entire `ResolveWithPortsLambda` function.

**Comment by eriknordmark** [comment]
> I guess we do not cache if resolving the name did not succeed and then it gets called again and again.

@milan-zededa 
Do we know that names we try to resolve using this code? Is it only the controller domain name? Or also things used by diag (like www.google.com)? 

If we know that we should be try to reproduce the issue by having DNS lookups fail for some of the names which we resolve (but still want the device to be able to talk to the controller).

**Review by eriknordmark — APPROVED** [review]
LGTM - re-run tests

**Comment by milan-zededa** [comment]
> > I guess we do not cache if resolving the name did not succeed and then it gets called again and again.
> 
> @milan-zededa Do we know that names we try to resolve using this code? Is it only the controller domain name? Or also things used by diag (like [www.google.com](http://www.google.com))?
> 
> If we know that we should be try to reproduce the issue by having DNS lookups fail for some of the names which we resolve (but still want the device to be able to talk to the controller).

It is only used for the controller domain name.
Maybe we could reproduce this issue by having multiple DNS servers published by the DHCP server, where one can resolve controller name while others fail (or possibly are not reachable/responsive).

**Review by OhmSpectator — APPROVED** [review]
I think we should merge now and backport into 13.4-stable, 12.0-stable and 11.0-stable...

**Comment by OhmSpectator** [comment]
@christoph-zededa, in the description, I see that it fixes a bug introduced in the 9.* branch. And I see the "stable" label. Nevertheless, I don't see any backport. I would expect it to be at least a part of 13.4-stable at that time. So, don't we need it? If so, I will remove the label.

**Comment by OhmSpectator** [comment]
> > @christoph-zededa, in the description, I see that it fixes a bug introduced in the 9.* branch. And I see the "stable" label. Nevertheless, I don't see any backport. I would expect it to be at least a part of 13.4-stable at that time. So, don't we need it? If so, I will remove the label.
> 
> here it is: #4425

Thanks! It lacked the mention of the original PR, so I missed it.

**Comment by christoph-zededa** [comment]
> > > @christoph-zededa, in the description, I see that it fixes a bug introduced in the 9.* branch. And I see the "stable" label. Nevertheless, I don't see any backport. I would expect it to be at least a part of 13.4-stable at that time. So, don't we need it? If so, I will remove the label.
> > 
> > 
> > here it is: #4425
> 
> Thanks! It lacked the mention of the original PR, so I missed it.

perhaps you can add something like this to your script:
```
#!/usr/bin/env perl

use strict;
use warnings;

use JSON 'decode_json';

my $cmd;
my $j;

my $pr = shift or die "Usage: $0 <PR>";

open $cmd, "-|", "gh pr view ".$pr." --json title";
$j = decode_json <$cmd>;

my $pr_title = $j->{'title'};
open $cmd, "-|", "gh pr list  --state all --search '".$pr_title."' --json title,number ";

$j = decode_json <$cmd>;

foreach my $pr (@$j) {
	if ($pr->{'title'} eq $pr_title) {
		next;
	}
	if ($pr->{'title'} =~ m/^\[[^\]]+\] $pr_title/) {
		print $pr->{'number'}."\n";
	}
}
```

**Comment by OhmSpectator** [comment]
> > > > @christoph-zededa, in the description, I see that it fixes a bug introduced in the 9.* branch. And I see the "stable" label. Nevertheless, I don't see any backport. I would expect it to be at least a part of 13.4-stable at that time. So, don't we need it? If so, I will remove the label.
> > > 
> > > 
> > > here it is: #4425
> > 
> > 
> > Thanks! It lacked the mention of the original PR, so I missed it.
> 
> perhaps you can add something like this to your script:
> 
> ```
> #!/usr/bin/env perl
> 
> use strict;
> use warnings;
> 
> use JSON 'decode_json';
> 
> my $cmd;
> my $j;
> 
> my $pr = shift or die "Usage: $0 <PR>";
> 
> open $cmd, "-|", "gh pr view ".$pr." --json title";
> $j = decode_json <$cmd>;
> 
> my $pr_title = $j->{'title'};
> open $cmd, "-|", "gh pr list  --state all --search '".$pr_title."' --json title,number ";
> 
> $j = decode_json <$cmd>;
> 
> foreach my $pr (@$j) {
> 	if ($pr->{'title'} eq $pr_title) {
> 		next;
> 	}
> 	if ($pr->{'title'} =~ m/^\[[^\]]+\] $pr_title/) {
> 		print $pr->{'number'}."\n";
> 	}
> }
> ```

My script also handles the commit messages added by `-x` in the commits, but it is still not perfect. That's why I have to reverify everything manually.

**Pull request merged**
